File: apps/badminton-coach/web/js/core/biomech.js
/**
 * Turn a frame of pose landmarks into badminton-relevant joint angles.
 *
 * ## Which landmarks
 *
 * Angles are computed from MediaPipe's *world* landmarks, which are metric
 * (metres), origin at the mid-hip, and independent of where the player is in the
 * frame. Image landmarks would make the same joint read differently depending on
 * whether the player is near or far from the phone, which is exactly the
 * situation on a badminton court.
 *
 * ## Coordinate convention
 *
 * MediaPipe world landmarks use y-down and z-away-from-camera. {@link toUpFrame}
 * flips both, giving a right-handed frame with **y up** and **z towards the
 * camera**, which is what the rest of this file assumes.
 *
 * ## The body frame
 *
 * Angles that describe *technique* rather than *anatomy* -- is the arm across the
 * body? is the chest turned? -- only mean something relative to the player. So we
 * build an orthonormal frame on the torso (`right`, `up`, `forward`) and express
 * the racket arm in it. That makes every measurement invariant to which way the
 * player happens to be facing the phone.
 *
 * One caveat worth stating plainly: MediaPipe's world frame is aligned to the
 * *camera*, not to gravity. Angles between body parts (elbow, knee, arm-to-trunk,
 * shoulder-to-hip twist) are unaffected. Angles quoted against the vertical --
 * trunk lean -- assume the phone is held roughly upright and level, and are
 * flagged as such in the output.
 */

import {
  sub, add, cross, dot, normalize, length, distance, midpoint,
  reject, angleBetween, jointAngle, signedAngle,
} from './vec3.js';
import { LM, side, otherArm } from './landmarks.js';

/** Convert one MediaPipe world landmark to the y-up, z-toward-camera frame. */
export const toUpFrame = (p) => [p[0], -p[1], -p[2]];

/** Convert a whole landmark array. */
export const worldToUpFrame = (points) => points.map(toUpFrame);

/**
 * The torso frame: an orthonormal basis attached to the player.
 *
 * - `right` points out of the player's right shoulder,
 * - `up` runs mid-hip to mid-shoulder,
 * - `forward` comes out of the chest.
 *
 * `up` is taken as primary and `right` is orthogonalised against it, because the
 * shoulder line is the noisier of the two when an arm is swinging.
 */
export function bodyFrame(world) {
  const midShoulder = midpoint(world[LM.LEFT_SHOULDER], world[LM.RIGHT_SHOULDER]);
  const midHip = midpoint(world[LM.LEFT_HIP], world[LM.RIGHT_HIP]);
  const up = normalize(sub(midShoulder, midHip));
  const shoulderAxis = sub(world[LM.RIGHT_SHOULDER], world[LM.LEFT_SHOULDER]);
  const right = normalize(reject(shoulderAxis, up));
  // cross(up, right) points out of the chest in a right-handed y-up frame.
  const forward = normalize(cross(up, right));
  return { origin: midShoulder, midShoulder, midHip, right, up, forward };
}

/** Express a world point in the body frame, in metres. */
export function toBody(frame, point) {
  const d = sub(point, frame.origin);
  return [dot(d, frame.right), dot(d, frame.up), dot(d, frame.forward)];
}

/**
 * Body-scale estimates used to normalise lengths.
 *
 * Normalising by the player's own trunk makes thresholds transferable between a
 * tall adult and a junior, and between a clip shot from 5 m and one from 15 m.
 */
export function bodyScale(world) {
  const midShoulder = midpoint(world[LM.LEFT_SHOULDER], world[LM.RIGHT_SHOULDER]);
  const midHip = midpoint(world[LM.LEFT_HIP], world[LM.RIGHT_HIP]);
  return {
    trunk: distance(midShoulder, midHip),
    shoulderWidth: distance(world[LM.LEFT_SHOULDER], world[LM.RIGHT_SHOULDER]),
    hipWidth: distance(world[LM.LEFT_HIP], world[LM.RIGHT_HIP]),
    armLeft:
      distance(world[LM.LEFT_SHOULDER], world[LM.LEFT_ELBOW]) +
      distance(world[LM.LEFT_ELBOW], world[LM.LEFT_WRIST]),
    armRight:
      distance(world[LM.RIGHT_SHOULDER], world[LM.RIGHT_ELBOW]) +
      distance(world[LM.RIGHT_ELBOW], world[LM.RIGHT_WRIST]),
  };
}

/** Elbow flexion: 180 is a straight arm, 90 a right angle. */
export function elbowAngle(world, arm) {
  return jointAngle(
    world[side('SHOULDER', arm)],
    world[side('ELBOW', arm)],
    world[side('WRIST', arm)],
  );
}

/** Knee flexion: 180 is a straight leg. */
export function kneeAngle(world, leg) {
  return jointAngle(
    world[side('HIP', leg)],
    world[side('KNEE', leg)],
    world[side('ANKLE', leg)],
  );
}

/**
 * How far the upper arm is raised away from the side of the body, in degrees.
 *
 * 0 is the arm hanging down the flank; 90 is horizontal; 180 is straight
 * overhead. This is the number a coach means by "get your elbow up".
 */
export function shoulderElevation(frame, world, arm) {
  const upperArm = sub(world[side('ELBOW', arm)], world[side('SHOULDER', arm)]);
  const down = [-frame.up[0], -frame.up[1], -frame.up[2]];
  return angleBetween(upperArm, down);
}

/**
 * Where the upper arm points in the horizontal plane, in degrees.
 *
 * Measured from straight ahead of the chest and signed towards the racket side:
 * near 0 the arm is out in front, +90 is straight out to the racket side, and a
 * negative value means the arm has crossed the body -- the geometry of a
 * backhand.
 */
export function shoulderAzimuth(frame, world, arm, racketArm = arm) {
  const upperArm = sub(world[side('ELBOW', arm)], world[side('SHOULDER', arm)]);
  const flat = reject(upperArm, frame.up);
  const raw = signedAngle(frame.forward, flat, frame.up);
  // signedAngle is positive towards the player's left about `up`; flip so that
  // positive always means "towards the racket side" for either handedness.
  const towardsRacketSide = racketArm === 'right' ? -1 : 1;
  return raw * towardsRacketSide;
}

/**
 * Trunk lean away from vertical, decomposed into forwards and sideways.
 *
 * `total` is the tilt of the mid-hip-to-mid-shoulder line away from vertical.
 * `forward` is positive when leaning the way the player is facing, `lateral`
 * positive when leaning towards the racket side.
 *
 * The decomposition is taken against *world-horizontal* axes derived from which
 * way the player faces -- not against the torso's own axes, which are orthogonal
 * to its up-vector by construction and would make every component zero.
 *
 * Vertical here is the camera's vertical, so this is the one measurement in the
 * file that assumes the phone is held roughly upright.
 */
export function trunkLean(frame, racketArm = 'right') {
  const up = frame.up;
  const vertical = [0, 1, 0];
  const total = angleBetween(up, vertical);

  // Where the player faces, flattened onto the floor plane.
  let heading = normalize(reject(frame.forward, vertical));
  if (length(heading) < 1e-6) {
    // Chest pointing straight up or down (a deep dive, or a bad frame): fall back
    // to the shoulder line so the axes stay defined.
    heading = normalize(reject(frame.right, vertical));
  }
  // With y up, cross(heading, vertical) points out of the player's right side.
  const rightAxis = normalize(cross(heading, vertical));
  const towardsRacketSide = racketArm === 'right' ? 1 : -1;

  const upright = dot(up, vertical);
  const forward = Math.atan2(dot(up, heading), upright) * (180 / Math.PI);
  const lateral =
    Math.atan2(dot(up, rightAxis), upright) * (180 / Math.PI) * towardsRacketSide;
  return { total, forward, lateral, heading };
}

/**
 * Shoulder-hip separation ("X-factor"), in degrees.
 *
 * The twist stored between the pelvis and the ribcage during the backswing is
 * where an overhead shot's power comes from; unwinding it is what a coach means
 * by "hit with your body, not your arm". Positive means the shoulders are turned
 * further away from the target than the hips -- the loaded direction for the
 * given racket hand.
 */
export function hipShoulderSeparation(frame, world, racketArm = 'right') {
  const shoulderLine = reject(
    sub(world[LM.RIGHT_SHOULDER], world[LM.LEFT_SHOULDER]),
    frame.up,
  );
  const hipLine = reject(sub(world[LM.RIGHT_HIP], world[LM.LEFT_HIP]), frame.up);
  const raw = signedAngle(hipLine, shoulderLine, frame.up);
  return raw * (racketArm === 'right' ? 1 : -1);
}

/**
 * Stance width as a multiple of shoulder width.
 *
 * Around 1 is a normal standing base; a badminton ready stance is wider, and a
 * lunge to the net is wider still.
 */
export function stanceWidth(world, scale) {
  const feet = distance(world[LM.LEFT_ANKLE], world[LM.RIGHT_ANKLE]);
  return scale.shoulderWidth > 1e-6 ? feet / scale.shoulderWidth : NaN;
}

/**
 * Everything about the racket hand's position, in body-frame trunk lengths.
 *
 * - `height`: 0 is level with the shoulders, positive is above them.
 * - `lateral`: positive on the racket side, negative once the hand has crossed
 *   the midline. This single number is what separates a forehand from a
 *   backhand.
 * - `forward`: positive in front of the chest.
 * - `extension`: 0..1, how much of the arm's length is actually being used.
 */
export function handPosition(frame, world, arm, trunk, armLength, racketArm = arm) {
  const wrist = toBody(frame, world[side('WRIST', arm)]);
  const unit = trunk > 1e-6 ? trunk : NaN;
  const towardsRacketSide = racketArm === 'right' ? 1 : -1;
  const reach = distance(world[side('WRIST', arm)], world[side('SHOULDER', arm)]);
  return {
    height: wrist[1] / unit,
    lateral: (wrist[0] * towardsRacketSide) / unit,
    forward: wrist[2] / unit,
    extension: armLength > 1e-6 ? Math.min(1.5, reach / armLength) : NaN,
    body: wrist,
  };
}

/** Mean visibility of the landmarks that the angles above depend on. */
export function coreVisibility(image, indices) {
  let sum = 0;
  let n = 0;
  for (const i of indices) {
    const v = image?.[i]?.[3];
    if (Number.isFinite(v)) {
      sum += v;
      n += 1;
    }
  }
  return n ? sum / n : 0;
}

/**
 * Compute the full metric set for one frame.
 *
 * @param {number[][]} world world landmarks, already in the y-up frame
 * @param {object} options
 * @param {'left'|'right'} options.racketArm which hand holds the racket
 * @param {number} options.trunk stable trunk length in metres (a running median,
 *   not this frame's value -- see `filters.RunningMedian`)
 * @param {number} options.armLength stable racket-arm length in metres
 */
export function frameMetrics(world, { racketArm = 'right', trunk, armLength } = {}) {
  const frame = bodyFrame(world);
  const scale = bodyScale(world);
  const trunkLen = Number.isFinite(trunk) && trunk > 1e-6 ? trunk : scale.trunk;
  const armLen =
    Number.isFinite(armLength) && armLength > 1e-6
      ? armLength
      : racketArm === 'right'
        ? scale.armRight
        : scale.armLeft;
  const off = otherArm(racketArm);

  return {
    frame,
    scale,
    racketArm,
    elbow: elbowAngle(world, racketArm),
    elbowOff: elbowAngle(world, off),
    shoulderElevation: shoulderElevation(frame, world, racketArm),
    shoulderElevationOff: shoulderElevation(frame, world, off),
    shoulderAzimuth: shoulderAzimuth(frame, world, racketArm, racketArm),
    trunkLean: trunkLean(frame, racketArm),
    separation: hipShoulderSeparation(frame, world, racketArm),
    kneeLeft: kneeAngle(world, 'left'),
    kneeRight: kneeAngle(world, 'right'),
    stanceWidth: stanceWidth(world, scale),
    hand: handPosition(frame, world, racketArm, trunkLen, armLen, racketArm),
    trunkLength: trunkLen,
    armLength: armLen,
  };
}

File: apps/badminton-coach/analyze/badminton_coach/biomech.py
"""Turn a frame of pose landmarks into badminton-relevant joint angles.

Mirrors ``web/js/core/biomech.js``, which carries the full rationale. In brief:

* Angles come from MediaPipe's *world* landmarks -- metric, hip-centred, and
  independent of where in the frame the player is, which matters when the player
  runs from the net to the back of the court.
* :func:`to_up_frame` flips MediaPipe's y-down, z-away convention into a
  right-handed frame with **y up** and **z towards the camera**.
* Technique angles are expressed in a frame attached to the torso, so they do not
  change meaning when the player turns to face a different way.
* One caveat: MediaPipe's world frame is aligned to the camera, not to gravity.
  Angles between body parts are unaffected; trunk lean, which is quoted against
  vertical, assumes the phone is held roughly upright.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from . import landmarks as L
from .geometry import (
    angle_between, cross, distance, dot, joint_angle, length, midpoint,
    normalize, reject, signed_angle, sub,
)

NAN = float("nan")


def to_up_frame(p):
    """One MediaPipe world landmark, in the y-up, z-towards-camera frame."""
    return (p[0], -p[1], -p[2])


def world_to_up_frame(points):
    return [to_up_frame(p) for p in points]


@dataclass(frozen=True)
class BodyFrame:
    """An orthonormal basis attached to the player.

    ``right`` points out of the player's right shoulder, ``up`` runs mid-hip to
    mid-shoulder, ``forward`` comes out of the chest. ``up`` is taken as primary
    and ``right`` orthogonalised against it, because the shoulder line is the
    noisier of the two while an arm is swinging.
    """

    origin: tuple
    mid_shoulder: tuple
    mid_hip: tuple
    right: tuple
    up: tuple
    forward: tuple


def body_frame(world) -> BodyFrame:
    mid_shoulder = midpoint(world[L.LEFT_SHOULDER], world[L.RIGHT_SHOULDER])
    mid_hip = midpoint(world[L.LEFT_HIP], world[L.RIGHT_HIP])
    up = normalize(sub(mid_shoulder, mid_hip))
    shoulder_axis = sub(world[L.RIGHT_SHOULDER], world[L.LEFT_SHOULDER])
    right = normalize(reject(shoulder_axis, up))
    # cross(up, right) points out of the chest in a right-handed y-up frame.
    forward = normalize(cross(up, right))
    return BodyFrame(mid_shoulder, mid_shoulder, mid_hip, right, up, forward)


def to_body(frame: BodyFrame, point):
    """Express a world point in the body frame, in metres."""
    d = sub(point, frame.origin)
    return (dot(d, frame.right), dot(d, frame.up), dot(d, frame.forward))


@dataclass(frozen=True)
class BodyScale:
    trunk: float
    shoulder_width: float
    hip_width: float
    arm_left: float
    arm_right: float


def body_scale(world) -> BodyScale:
    """Body-scale estimates used to normalise lengths.

    Normalising by the player's own trunk makes thresholds transferable between a
    tall adult and a junior, and between a clip shot from 5 m and one from 15 m.
    """
    mid_shoulder = midpoint(world[L.LEFT_SHOULDER], world[L.RIGHT_SHOULDER])
    mid_hip = midpoint(world[L.LEFT_HIP], world[L.RIGHT_HIP])
    return BodyScale(
        trunk=distance(mid_shoulder, mid_hip),
        shoulder_width=distance(world[L.LEFT_SHOULDER], world[L.RIGHT_SHOULDER]),
        hip_width=distance(world[L.LEFT_HIP], world[L.RIGHT_HIP]),
        arm_left=distance(world[L.LEFT_SHOULDER], world[L.LEFT_ELBOW])
        + distance(world[L.LEFT_ELBOW], world[L.LEFT_WRIST]),
        arm_right=distance(world[L.RIGHT_SHOULDER], world[L.RIGHT_ELBOW])
        + distance(world[L.RIGHT_ELBOW], world[L.RIGHT_WRIST]),
    )


def elbow_angle(world, arm: str) -> float:
    """Elbow flexion: 180 is a straight arm."""
    return joint_angle(
        world[L.side("SHOULDER", arm)],
        world[L.side("ELBOW", arm)],
        world[L.side("WRIST", arm)],
    )


def knee_angle(world, leg: str) -> float:
    """Knee flexion: 180 is a straight leg."""
    return joint_angle(
        world[L.side("HIP", leg)],
        world[L.side("KNEE", leg)],
        world[L.side("ANKLE", leg)],
    )


def shoulder_elevation(frame: BodyFrame, world, arm: str) -> float:
    """How far the upper arm is raised from the side of the body, in degrees.

    0 is the arm hanging down the flank, 90 horizontal, 180 straight overhead --
    the number a coach means by "get your elbow up".
    """
    upper_arm = sub(world[L.side("ELBOW", arm)], world[L.side("SHOULDER", arm)])
    down = (-frame.up[0], -frame.up[1], -frame.up[2])
    return angle_between(upper_arm, down)


def shoulder_azimuth(frame: BodyFrame, world, arm: str, racket_arm: str | None = None) -> float:
    """Where the upper arm points in the horizontal plane, in degrees.

    Measured from straight ahead of the chest, signed towards the racket side:
    near 0 the arm is in front, +90 straight out to the racket side, negative once
    the arm has crossed the body -- the geometry of a backhand.
    """
    racket_arm = racket_arm or arm
    upper_arm = sub(world[L.side("ELBOW", arm)], world[L.side("SHOULDER", arm)])
    flat = reject(upper_arm, frame.up)
    raw = signed_angle(frame.forward, flat, frame.up)
    # signed_angle is positive towards the player's left about `up`; flip so that
    # positive always means "towards the racket side" for either handedness.
    return raw * (-1 if racket_arm == "right" else 1)


def trunk_lean(frame: BodyFrame, racket_arm: str = "right") -> dict:
    """Trunk lean away from vertical, split into forwards and sideways.

    The decomposition is taken against *world-horizontal* axes derived from which
    way the player faces -- not against the torso's own axes, which are orthogonal
    to its up-vector by construction and would make every component zero.
    """
    up = frame.up
    vertical = (0.0, 1.0, 0.0)
    total = angle_between(up, vertical)

    heading = normalize(reject(frame.forward, vertical))
    if length(heading) < 1e-6:
        # Chest pointing straight up or down (a dive, or a bad frame): fall back
        # to the shoulder line so the axes stay defined.
        heading = normalize(reject(frame.right, vertical))
    # With y up, cross(heading, vertical) points out of the player's right side.
    right_axis = normalize(cross(heading, vertical))

    upright = dot(up, vertical)
    forward = math.degrees(math.atan2(dot(up, heading), upright))
    lateral = math.degrees(math.atan2(dot(up, right_axis), upright))
    return {
        "total": total,
        "forward": forward,
        "lateral": lateral * (1 if racket_arm == "right" else -1),
        "heading": heading,
    }


def hip_shoulder_separation(frame: BodyFrame, world, racket_arm: str = "right") -> float:
    """Shoulder-hip separation ("X-factor"), in degrees.

    The twist stored between pelvis and ribcage during the backswing is where an
    overhead's power comes from; unwinding it is what a coach means by "hit with
    your body, not your arm".
    """
    shoulder_line = reject(sub(world[L.RIGHT_SHOULDER], world[L.LEFT_SHOULDER]), frame.up)
    hip_line = reject(sub(world[L.RIGHT_HIP], world[L.LEFT_HIP]), frame.up)
    raw = signed_angle(hip_line, shoulder_line, frame.up)
    return raw * (1 if racket_arm == "right" else -1)


def stance_width(world, scale: BodyScale) -> float:
    """Stance width as a multiple of shoulder width."""
    feet = distance(world[L.LEFT_ANKLE], world[L.RIGHT_ANKLE])
    return feet / scale.shoulder_width if scale.shoulder_width > 1e-6 else NAN


def hand_position(frame: BodyFrame, world, arm: str, trunk: float,
                  arm_length: float, racket_arm: str | None = None) -> dict:
    """The racket hand's position, in body-frame trunk lengths.

    ``height`` is 0 at the shoulder line and positive above it; ``lateral`` is
    positive on the racket side and negative once the hand has crossed the
    midline, which is the single number separating forehand from backhand;
    ``forward`` is positive in front of the chest; ``extension`` is how much of
    the arm's length is being used, 0..1.
    """
    racket_arm = racket_arm or arm
    wrist = to_body(frame, world[L.side("WRIST", arm)])
    unit = trunk if trunk > 1e-6 else NAN
    towards_racket_side = 1 if racket_arm == "right" else -1
    reach = distance(world[L.side("WRIST", arm)], world[L.side("SHOULDER", arm)])
    return {
        "height": wrist[1] / unit,
        "lateral": (wrist[0] * towards_racket_side) / unit,
        "forward": wrist[2] / unit,
        "extension": min(1.5, reach / arm_length) if arm_length > 1e-6 else NAN,
        "body": wrist,
    }


def core_visibility(image, indices=None) -> float:
    """Mean visibility of the landmarks the angles depend on."""
    indices = L.CORE if indices is None else indices
    values = [image[i][3] for i in indices if len(image[i]) > 3]
    values = [v for v in values if v is not None and math.isfinite(v)]
    return sum(values) / len(values) if values else 0.0


def frame_metrics(world, racket_arm: str = "right", trunk: float | None = None,
                  arm_length: float | None = None) -> dict:
    """The full metric set for one frame.

    ``trunk`` and ``arm_length`` should be stable estimates (a running median),
    not this frame's raw values.
    """
    frame = body_frame(world)
    scale = body_scale(world)
    trunk_len = trunk if (trunk is not None and math.isfinite(trunk) and trunk > 1e-6) else scale.trunk
    if arm_length is not None and math.isfinite(arm_length) and arm_length > 1e-6:
        arm_len = arm_length
    else:
        arm_len = scale.arm_right if racket_arm == "right" else scale.arm_left
    off = L.other_arm(racket_arm)

    return {
        "frame": frame,
        "scale": scale,
        "racket_arm": racket_arm,
        "elbow": elbow_angle(world, racket_arm),
        "elbow_off": elbow_angle(world, off),
        "shoulder_elevation": shoulder_elevation(frame, world, racket_arm),
        "shoulder_elevation_off": shoulder_elevation(frame, world, off),
        "shoulder_azimuth": shoulder_azimuth(frame, world, racket_arm, racket_arm),
        "trunk_lean": trunk_lean(frame, racket_arm),
        "separation": hip_shoulder_separation(frame, world, racket_arm),
        "knee_left": knee_angle(world, "left"),
        "knee_right": knee_angle(world, "right"),
        "stance_width": stance_width(world, scale),
        "hand": hand_position(frame, world, racket_arm, trunk_len, arm_len, racket_arm),
        "trunk_length": trunk_len,
        "arm_length": arm_len,
    }
